File: frontend/swift-objectiveC/knownLanguages/knownLanguages/ProfileVC.m
//
//  ProfileVC.m
//  knownLanguages
//
//  Created by Tan Junhe on 20/1/24.
//

#import "ProfileVC.h"
#import "knownLanguages-Swift.h"
 
@interface ProfileVC ()

@property (weak, nonatomic) IBOutlet UITextField *nameTF;
@property (weak, nonatomic) IBOutlet UITextField *currentPWTF;
@property (weak, nonatomic) IBOutlet UITextField *PWTFnew;
@property (weak, nonatomic) IBOutlet UITextField *emailTF;
@property (weak, nonatomic) IBOutlet UITextField *dobTF;
@property (weak, nonatomic) IBOutlet UIPickerView *genderPV;

@property UIDatePicker *datePicker;

@property NSString *url;
@property NSArray *genderList;

@property NSString *name;
@property NSString *token;
@property NSString *dob;
@property NSString *gender;

@property FormatDate *FD;

@end

@implementation ProfileVC
- (void)viewDidLoad {
    [super viewDidLoad];
    [self startLoad];
    
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    [self startLoad];
    
}

- (void)startLoad{
    _url = [[IPSetting sharedInstance] setUrl];

    _genderList = [UserGender sharedInstance].gender;
    
    _FD = [[FormatDate alloc] init];
    
    _name = [[NSUserDefaults standardUserDefaults] stringForKey:@"name"];
    _token = [[NSUserDefaults standardUserDefaults] stringForKey:@"token"];
    
    NSDictionary *jsonBodyDict = @{@"name": _name, @"token":_token};
    NSData *params = [NSJSONSerialization dataWithJSONObject:jsonBodyDict options:kNilOptions error:nil];
    [self apiCall:@"profile" :params :@"POST"];
    
    [_nameTF setText:_name];
    
    _datePicker = [[UIDatePicker alloc] init];
    _datePicker.datePickerMode = UIDatePickerModeDate;
    [_datePicker addTarget:self action:@selector(dateChange:) forControlEvents:UIControlEventValueChanged];
    _datePicker.frame = CGRectMake(0, 0, 0, 300);
    _datePicker.maximumDate = [[NSDate alloc] init];
    _dobTF.inputView = _datePicker;
    if (@available(iOS 14.0, *)) {
        _datePicker.preferredDatePickerStyle = UIDatePickerStyleInline;
    } else if (@available(iOS 13.4, *)) {
        _datePicker.preferredDatePickerStyle = UIDatePickerStyleWheels;
    } else {
        
    }
    _genderPV.delegate = self;
    _genderPV.dataSource = self;
    
}

- (IBAction)updateBTN:(id)sender {
    _gender = _gender == _genderList[0] ? @"" : _gender;
    NSDictionary *jsonBodyDict = @{@"name": _name,
                                   @"password": [[self currentPWTF] text],
                                   @"newPassword": [[self PWTFnew] text],
                                   @"email": [[self emailTF] text],
                                   @"dob": _dob,
                                   @"gender": _gender,
                                   @"token":_token};
    NSData *params = [NSJSONSerialization dataWithJSONObject:jsonBodyDict options:kNilOptions error:nil];
    [self apiCall:@"updateProfile" :params :@"PATCH"];
}

- (IBAction)logoutBTN:(id)sender {
    [self logout];
}

- (void)dateChange:(UIDatePicker *)datePicker {
    _dobTF.text = [_FD formatDateWithDate:datePicker.date];
    _dob = [_FD formatDateSQLWithDate:datePicker.date];
}


//pickerview gender
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
   return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return _genderList.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return _genderList[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    self.gender = _genderList[row];
}

- (void)logout{
    NSDictionary *jsonBodyDict = @{@"name": _name, @"token":_token};
    NSData *params = [NSJSONSerialization dataWithJSONObject:jsonBodyDict options:kNilOptions error:nil];
    [self apiCall:@"logout" :params :@"DELETE"];
}

- (void)alertShowUp: (NSString*)title : (NSString*)message{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];

    UIAlertAction *action = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                            //button click event
                        }];
    [alert addAction:action];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)apiCall: (NSString*)endPoint : (NSData*)params : (NSString*)resMethod{
    
    NSMutableURLRequest *request = [NSMutableURLRequest new];
    request.HTTPMethod = resMethod;
    
    [request setURL:[NSURL URLWithString: [NSString stringWithFormat:@"%@%@",_url,endPoint]]];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
    [request setHTTPBody:params];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if(error){
            dispatch_async(dispatch_get_main_queue(), ^{
                [self alertShowUp:@"Error" :[error localizedDescription]];
            });
        }else{
            NSError *error;
            NSDictionary *jsonResult = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
            if(!error) {
                if([endPoint isEqualToString:@"logout"]){
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [[EndSession sharedInstance] logout];
                    });
                }else if([endPoint isEqualToString:@"profile"]){
                    
                    
                    NSArray<NSDictionary *> *result = (NSArray<NSDictionary *> *)jsonResult[@"result"];
                    NSDictionary *result0 = result[0];
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [[self emailTF] setText:[result0 objectForKey:@"email"]];

                        self.dob = [[result0 objectForKey:@"DOB"] isKindOfClass:[NSString class]] ? [self.FD formatDateWithDate:[self.FD dateFromISOStringWithDateString:[result0 objectForKey:@"DOB"]]] : @"";
                    
                        [self.dobTF setText:self.dob];

                        self.datePicker.date = self.dob ? [self.FD dateForPVWithDate:[self.FD formatDateWithDate:[self.FD dateFromISOStringWithDateString:self.dob]]] : [NSDate date];
                        
                        self.gender = [[result0 objectForKey:@"gender"] isKindOfClass:[NSString class]] ? [result0 objectForKey:@"gender"] : self.genderList[0];
                        
                        [self.genderPV selectRow:[self.genderList indexOfObject:self.gender] inComponent:0 animated:YES];
                    
                    });
                    
                }else if([endPoint isEqualToString:@"updateProfile"]){
                    bool check = [[jsonResult objectForKey:@"check"] boolValue];
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self alertShowUp:check ? @"Successfully" : @"Error" :[jsonResult objectForKey:@"result"]];
                        [[self currentPWTF] setText:@""];
                    });
                }
            }else{
                NSLog(@"%@",error);
            }

        }
        
    }];
    [dataTask resume];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self.view endEditing:YES];
}
@end

File: frontend/swift/knownLanguages/knownLanguages/ProfileVC.swift
//
//  ProfileVC.swift
//  knownLanguages
//
//  Created by Tan Junhe on 29/7/23.
//

import UIKit

class ProfileVC: UIViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    
    @IBOutlet weak var nameTF: UITextField!
    @IBOutlet weak var currentPWTF: UITextField!
    @IBOutlet weak var newPWFT: UITextField!
    @IBOutlet weak var emailTF: UITextField!
    @IBOutlet weak var dobTF: UITextField!
    @IBOutlet weak var genderPV: UIPickerView!
    let datePicker = UIDatePicker()

    var FD = FormatDate();
    
    var url = IPSetting.sharedInstance.setUrl()
    var genderList = UserGender.shareInstance.gender
    
    var name = UserDefaults.standard.string(forKey: "name")
    var token = UserDefaults.standard.string(forKey: "token")
    var dob = ""
    var gender = ""
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        startload()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        startload()
    }
    
    func startload(){
        //get user info
        let params = ["name":UserDefaults.standard.string(forKey: "name"),
                      "token":UserDefaults.standard.string(forKey: "token")]
        self.apiCall(endPoint: "profile", params: params, resMethod: "POST")
        
        nameTF.text = name
        
        datePicker.datePickerMode = .date
        datePicker.addTarget(self, action: #selector(dateChange(datePicker:)), for: UIControl.Event.valueChanged)
        datePicker.frame.size = CGSize(width: 0, height: 300)
        datePicker.maximumDate = Date()
        dobTF.inputView = datePicker
        if #available(iOS 14.0, *) {
            datePicker.preferredDatePickerStyle = .inline
        } else {
            if #available(iOS 13.4, *) {
                datePicker.preferredDatePickerStyle = .wheels
            } else {
                
            }
        }
        
        genderPV.delegate = self
        genderPV.dataSource = self
    }
    
    @IBAction func updateBtn(_ sender: Any) {
        gender = gender == genderList[0] ? "" : gender
        dob = dobTF.text!.isEmpty ? "" : FD.formatDateSQL(date: datePicker.date)
        let params = ["name":name,
                      "password":currentPWTF.text!,
                      "newPassword":newPWFT.text!,
                      "email":emailTF.text!,
                      "dob":dob,
                      "gender": gender,
                      "token":token]
        self.apiCall(endPoint: "updateProfile", params: params, resMethod: "PATCH")
    }
    
    @IBAction func logoutBtn(_ sender: Any) {
        logout()
    }
    
    //datePicker
    @objc func dateChange(datePicker: UIDatePicker){
        dobTF.text = FD.formatDate(date: datePicker.date)
        dob = FD.formatDateSQL(date: datePicker.date)
    }
    
    //pickerView gender
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return genderList.count
    }
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return genderList[row]
    }
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int)
    {
        gender = genderList[row]
    }
    
    func logout(){
        let params = ["name":name,
                      "token":token]
        self.apiCall(endPoint: "logout", params: params , resMethod: "DELETE")
    }
    
    func alertShowUp(title: String, message: String){
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let action = UIAlertAction(title: "OK", style: .default)
        {(action: UIAlertAction!) in
            
        }
        alert.addAction(action)
        self.present(alert, animated: true)
    }

    
    func apiCall(endPoint: String, params: Any, resMethod: String ){
        var request = URLRequest(url: URL(string: "\(url)\(endPoint)")!)
        request.httpMethod = resMethod
        request.httpBody = try? JSONSerialization.data(withJSONObject: params, options: [])
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        let session = URLSession.shared
        let task = session.dataTask(with: request, completionHandler: { (data, response, error) in
            if let error = error {
                DispatchQueue.main.async {
                    self.alertShowUp(title: "Error", message: error.localizedDescription)
                }
            }else{
                do {
                    let json = try JSONSerialization.jsonObject(with: data!) as! Dictionary<String, AnyObject>
                    if(endPoint == "logout"){
                        DispatchQueue.main.async {
                            EndSession.sharedInstance.logout()
                        }
                    }else if(endPoint == "profile"){
                        if((json["check"])! as! Bool){
                            let result = json["result"] as! Array<Dictionary<String, AnyObject>>
                            let result0 = result[0]
                            DispatchQueue.main.async {
                                self.emailTF.text = result0["email"]! as? String
                                
                                let resultdob = result0["DOB"]! as Any
                                self.dob = resultdob is String ? self.FD.formatDate(date: self.FD.dateFromISOString(dateString: resultdob as! String)) : ""
                                self.dobTF.text = self.dob
                                
                                self.datePicker.date = !self.dob.isEmpty ?
                                self.FD.dateForPV(date: self.dob) : Date()
                                
                                let resultgender = result0["gender"]! as Any
                                self.gender = resultgender is String ? resultgender as! String : self.genderList[0]
                                self.genderPV.selectRow(self.genderList.firstIndex(of: self.gender) ?? 0, inComponent: 0, animated: true)
                            }
                        }
                    }else if(endPoint == "updateProfile"){
                        DispatchQueue.main.async {
                            self.alertShowUp(title: json["check"]! as! Bool ? "Successfully" : "Error" , message: json["result"] as! String)
                            self.currentPWTF.text = ""
                        }
                    }
                }catch{
                    print(error)
                }
            }
        })
        task.resume()
    }
    
    //dismiss keyboard
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        view.endEditing(true)
    }
}
